README
README for Ruby wrapper of Net::Patricia
---------------------------------------

rPatricia - A ruby wrapper of Net::Patricia

DESCRIPTION
-----------

  This is a ruby wrapper of Net::Patricia, which has been developed by 
  Dave Plonka. The original Net::Patricia and its C API are available
  from:
  http://net.doit.wisc.edu/~plonka/Net-Patricia/

  Net::Patricia is a module for fast IP address/prefix lookups.
  I have modified some interfaces for the Ruby wrapper version. 

NO WARANTY
----------    
    THE PROGRAM IS DISTRIBUTED IN THE HOPE THAT IT WILL BE USEFUL, BUT
    WITHOUT ANY WARRANTY. IT IS PROVIDED "AS IS" WITHOUT WARRANTY OF
    ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED
    TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
    PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND
    PERFORMANCE OF THE PROGRAM IS WITH YOU. SHOULD THE PROGRAM PROVE
    DEFECTIVE, YOU ASSUME THE COST OF ALL NECESSARY SERVICING, REPAIR
    OR CORRECTION. 

INSTALLATION
------------

   This package extracts, builds, installs in the usual fashion, i.e.:

      $ tar xvfz rpatricia.tar.gz
      $ cd rpatricia
      $ ruby extconf.rb
      $ make
      $ ruby test.rb
      # make install

SYNOPSIS
--------

  require 'rpatricia'
  pt = Patricia.new
  pt.add("192.168.1.0/24")
  pt.add("127.0.0.0/8", "user_data")
  node = pt.search_best("127.0.0.1")
  puts node.data
  puts node.prefix
  puts node.network
  puts node.prefixlen
  pt.remove("127.0.0.0/8")
  puts pt.num_nodes
  pt.show_nodes
  pt.clear

METHODS
-------
new:

      pt = Patricia.new

    This is the class' constructor - it returns a Patricia object for
    handling IPv4 addresses.  To handle IPv6 addresses (only),
    Patricia.new may be called with an additional argument:

      pt = Patricia.new(:AF_INET6)

    The Patricia object will be destroyed automatically when there are
    no longer any references to it.

add:
      pt.add(key_string[,user_data])

    The first argument, key_string, is a network or subnet
    specification in canonical form, e.g. ``10.0.0.0/8'', where the
    number after the slash represents the number of bits in the
    netmask. If no mask width is specified, the longest possible mask
    is assumed, i.e. 32 bits for AF_INET addresses and 128 bits for
    AF_INET6 addresses.

    The second argument, user_data, is optional. If supplied, it
    should be a STRING object specifying the user data that will be 
    stored in the Patricia Trie node. Subsequently, this value will
    be returned by the match methods described below to indicate a 
    successful search. 

    If no second argument is passed, the key_string will be stored as
    the user data and therfore will likewise be returned by the match
    functions.

    On success, this method returns the object of the Patricia Trie 
    node.

family:

    Returns either :AF_INET or :AF_INET6 symbol depending on how the
    object was initialized.  A Patricia object may only handle IPv4 or
    IPv6 addresses.

add_node:    An alias of add.

search_best:

      pt.search_best(key_string);

    This method searches the Patricia Trie to find a matching node,
    according to normal subnetting rules for the address and mask
    specified.

    The key_string argument is a network or subnet specification in
    canonical form, e.g. ``10.0.0.0/8'', where the number after the
    slash represents the number of bits in the netmask. If no mask
    width value is specified, the longest mask is assumed, i.e. 32
    bits for AF_INET addresses and 128 bits for AF_INET6 addresses.

    If a matching node is found in the Patricia Trie, this method
    returns the object of the node. This method returns nil on
    failure.

match_best:    An alias of search_best.

search_exact:

      pt.search_exact(key_string);

    This method searches the Patricia Trie to find a matching
    node. Its semantics are exactly the same as those described for
    search_best except that the key must match a node exactly. I.e.,
    it is not sufficient that the address and mask specified merely
    falls within the subnet specified by a particular node.

match_exact:	An alias of search_exact.

include?:

      pt.include?(key_string);

    This method behaves like match_best, but returns true on success
    and false on failure.  This method is more efficient than match_best
    as it does not allocate a new object.

remove:

      pt.remove(key_string);

    This method removes the node which exactly matches the the address
    and mask specified from the Patricia Trie.

    If the matching node is found in the Patricia Trie, it is removed,
    and this method returns the true. This method returns false on
    failure. 

remove_node:	 An alias of remove

num_nodes:
       
       pt.num_nodes

     This method returns the number of nodes in the Patricia Trie.

show_nodes:

       pt.print_nodes

     This method prints all the nodes in the Patricia Trie.

data:

       node.data
   
     This method returns the user data of the Patricia Trie node.

network:

       node.network
   
     This method returns the network of the Patricia Trie node.

prefix:

       node.prefix
   
     This method returns the prefix of the Patricia Trie node.

prefixlen:

       node.prefixlen
   
     This method returns the prefix length of the Patricia Trie 
     node.

AUTHORS
-------
Tatsuya Mori <mori.tatsuya@gmail.com>
Eric Wong <normalperson@yhbt.net>


Changes
0.09 2011/10/20
 - added basic support for Enumerable mixin
 - fixed the Patricia#Node#network method to include the prefix length

0.08 2011/08/12
 - IPv6 support: Patricia.new(:AF_INET6)
 - minor internal cleanups and speedups

0.07 2010/09/23
 - fix segfault from unpredictable GC ordering
   when node objects outlive the trees they come from

0.06 2010/09/14
 - rely on Ruby GC for allocations/free
   no need to explicitly destroy objects anymore
 - Ruby 1.9.2 C API compatibility
 - allow creation of subclasses of Patricia class
 - non-String objects may be stored as node-data
 - Patricia::Node objects are returned when match succeeds
   It's no longer possible to call tree methods on node
   objects and cause segfaults.
 - Patricia#show_nodes may be given a custom IO-like object
 - sync docs with remove, match_best and match_exact behavior
 - ArgumentError is raised for invalid addresses
   no more assertion failures for bad addresses
 - Patricia#show_nodes and Patricia#num_nodes no longer
   segfaults on empty trees

0.05 2010/03/12
 - reorganized directory layout
   no chance of accidentally having 'extconf' in the require path

0.04 2008/2/19
 - fixed the bug of "clear" method (destroy)
   node data were not cleared correctly

 - fixed the memory leak in search methods
   I had to clear the prefix instances

0.03 2008/2/17
 - fixed the bug of allocating user_data strings

0.02 2008/1/31
 - fixed the bug of printing prefix for 64-bit architecture 
   from Aki Nakao

0.01 2007/11/16
 - the first version


test.rb
require 'rpatricia'

def my_test condition
  puts (condition)? "ok!" : "error!"
end

puts "test: creating Patricia object"
my_test(t = Patricia.new) 

puts "test: adding 127.0.0.0/24"
my_test(n = t.add("127.0.0.0/24"))

puts "  data of added node = #{n.data}"
puts "  prefix of added node = #{n.prefix}"
puts "  network of added node = #{n.network}"
puts "  prefixlen of added node = #{n.prefixlen}"

puts "adding 192.168.1.0/24, 192.168.2.0/24, 192.168.3.100"
t.add("192.168.1.0/24")
t.add("192.168.2.0/24")
t.add("192.168.3.100")

puts "test: match_best 127.0.0.1"
my_test(n = t.match_best("127.0.0.1"))

puts "test: match_exact 192.168.3.100"
my_test(n = t.match_best("192.168.3.100"))

puts "  data of matched node = #{n.data}"
puts "  prefix of matched node = #{n.prefix}"
puts "  network of matched node = #{n.network}"
puts "  prefixlen of matched node = #{n.prefixlen}"

puts "test: adding prefix 10.0.0.0/8 with user data of 'pref_10.0.0.0/8'"
my_test(n = t.add("10.0.0.0/8", "pref_10"))

puts "  data of added node = #{n.data}"
puts "  prefix of added node = #{n.prefix}"
puts "  network of added node = #{n.network}"
puts "  prefixlen of added node = #{n.prefixlen}"

puts "test: match string 10.0.0.1"
my_test(n = t.match_best("10.0.0.1"))

puts "  data of matched node = #{n.data}"
puts "  prefix of matched node = #{n.prefix}"
puts "  network of matched node = #{n.network}"
puts "  prefixlen of matched node = #{n.prefixlen}"

puts "test: remove '42.0.0.0/8'; should return nil"
my_test(!t.remove("42.0.0.0/8"))

puts "test: remove '10.0.0.0/8'"
my_test(n= t.remove("10.0.0.0/8"))


puts "test: number of nodes in the tree should be '4'"
my_test(4 == t.num_nodes)

puts "test: showing all nodes"
t.show_nodes

t.destroy


test/test_gc.rb
require 'test/unit'
require 'rpatricia'

# this takes a while, watch memory usage not grow in top(1) or similar

class TestGc < Test::Unit::TestCase

  def setup
    @arrays = Patricia.new
    @arrays.add('10.0.0.0/8', [])
    @arrays.add('127.0.0.0/24', [])

    @strings = Patricia.new
    @strings.add('10.0.0.0/8', "big lan")
    @strings.add('127.0.0.0/24', "localhost")
  end

  def test_clear
    1_000_000.times do
      @strings.clear
      assert_equal 0, @strings.num_nodes
      @strings.add('10.0.0.0/8', "big lan")
      assert_equal 1, @strings.num_nodes
      @strings.add('127.0.0.0/8', "localhost")
      assert_equal 2, @strings.num_nodes
      @strings.add('192.168.1.0/24', "home")
      assert_equal 3, @strings.num_nodes
    end
  end

  def test_gc_dup
    100000.times do
      tmp = @strings.dup
      tmp.remove '127.0.0.0/24'
      tmp = @arrays.dup
      tmp.remove '127.0.0.0/24'
    end
    assert_equal [], @arrays.match_best('127.0.0.1').data
    assert_equal "localhost", @strings.match_best('127.0.0.1').data
  end

  def test_gc
    assert_nothing_raised do
      500_000.times do
        t = Patricia.new
        t.add('10.0.0.0/8', {})
        t.add('127.0.0.0/24', "home sweet home")
      end
    end

    # ensure what we created originally didn't get GC-ed'
    500_000.times do
      assert_equal [], @arrays.match_best('127.0.0.1').data
      assert_equal "localhost", @strings.match_best('127.0.0.1').data
    end
  end

  def test_destroy
    assert @strings.destroy
    assert @strings.destroy
  end

end


test/test_ipv6.rb
require 'test/unit'
require 'rpatricia'

class TestIPv6 < Test::Unit::TestCase
  def test_ipv6
    t = Patricia.new :AF_INET6
    assert_equal :AF_INET6, t.family
    node = t.add('1234:567::/35', 'hello world')
    assert_kind_of Patricia::Node, node
    assert_equal 'hello world', node.data
    assert_equal 'hello world', t.match_best('1234:567::/128').data
  end

  def test_ipv4_mapped_ipv6
    t = Patricia.new :AF_INET6
    assert_equal :AF_INET6, t.family
    node = t.add('::ffff:10.0.0.0/8', 'hello world')
    assert_kind_of Patricia::Node, node
    assert_equal 'hello world', node.data
    assert_equal 'hello world', t.match_best('::ffff:10.0.0.1').data
  end
end


ext/rpatricia/patricia.h
/*
 * $Id: patricia.h,v 1.6 2005/12/07 20:53:01 dplonka Exp $
 * Dave Plonka <plonka@doit.wisc.edu>
 *
 * This product includes software developed by the University of Michigan,
 * Merit Network, Inc., and their contributors. 
 *
 * This file had been called "radix.h" in the MRT sources.
 *
 * I renamed it to "patricia.h" since it's not an implementation of a general
 * radix trie.  Also, pulled in various requirements from "mrt.h" and added
 * some other things it could be used as a standalone API.
 */

#ifndef _PATRICIA_H
#define _PATRICIA_H

/* typedef unsigned int u_int; */
typedef void (*void_fn_t)();
/* { from defs.h */
#define prefix_touchar(prefix) ((u_char *)&(prefix)->add.sin)
#define MAXLINE 1024
#define BIT_TEST(f, b)  ((f) & (b))
/* } */

#include <sys/types.h> /* for u_* definitions (on FreeBSD 5) */

#include <errno.h> /* for EAFNOSUPPORT */
#ifndef EAFNOSUPPORT
#  defined EAFNOSUPPORT WSAEAFNOSUPPORT
#  include <winsock.h>
#else
#  include <netinet/in.h> /* for struct in_addr */
#endif

#include <sys/socket.h> /* for AF_INET */

/* { from mrt.h */

typedef struct _prefix4_t {
    u_short family;		/* AF_INET | AF_INET6 */
    u_short bitlen;		/* same as mask? */
    int ref_count;		/* reference count */
    struct in_addr sin;
} prefix4_t;

typedef struct _prefix_t {
    u_short family;		/* AF_INET | AF_INET6 */
    u_short bitlen;		/* same as mask? */
    int ref_count;		/* reference count */
    union {
		struct in_addr sin;
		struct in6_addr sin6;
    } add;
} prefix_t;

/* } */

typedef struct _patricia_node_t {
   u_int bit;			/* flag if this node used */
   prefix_t *prefix;		/* who we are in patricia tree */
   struct _patricia_node_t *l, *r;	/* left and right children */
   struct _patricia_node_t *parent;/* may be used */
   void *data;			/* pointer to data */
   void	*user;		/* pointer to usr data (ex. route flap info) */
} patricia_node_t;

typedef struct _patricia_tree_t {
   patricia_node_t 	*head;
   u_int		maxbits;	/* for IP, 32 bit addresses */
   int num_active_node;		/* for debug purpose */
} patricia_tree_t;


patricia_node_t *patricia_search_exact (patricia_tree_t *patricia, prefix_t *prefix);
patricia_node_t *patricia_search_best (patricia_tree_t *patricia, prefix_t *prefix);
patricia_node_t * patricia_search_best2 (patricia_tree_t *patricia, prefix_t *prefix, 
				   int inclusive);
patricia_node_t *patricia_lookup (patricia_tree_t *patricia, prefix_t *prefix);
void patricia_remove (patricia_tree_t *patricia, patricia_node_t *node);
patricia_tree_t *New_Patricia (int maxbits);
void Clear_Patricia (patricia_tree_t *patricia, void_fn_t func);
void Destroy_Patricia (patricia_tree_t *patricia, void_fn_t func);
void patricia_process (patricia_tree_t *patricia, void_fn_t func);

/* { from demo.c */

prefix_t *
ascii2prefix (char *string, prefix_t *prefix);

/* } */

#define PATRICIA_MAXBITS 128
#define PATRICIA_MAXSTRLEN (INET6_ADDRSTRLEN + sizeof("/128") - 1)
#define PATRICIA_NBIT(x)        (0x80 >> ((x) & 0x7f))
#define PATRICIA_NBYTE(x)       ((x) >> 3)

#define PATRICIA_DATA_GET(node, type) (type *)((node)->data)
#define PATRICIA_DATA_SET(node, value) ((node)->data = (void *)(value))

#define PATRICIA_WALK(Xhead, Xnode) \
    do { \
        patricia_node_t *Xstack[PATRICIA_MAXBITS+1]; \
        patricia_node_t **Xsp = Xstack; \
        patricia_node_t *Xrn = (Xhead); \
        while ((Xnode = Xrn)) { \
            if (Xnode->prefix)

#define PATRICIA_WALK_ALL(Xhead, Xnode) \
do { \
        patricia_node_t *Xstack[PATRICIA_MAXBITS+1]; \
        patricia_node_t **Xsp = Xstack; \
        patricia_node_t *Xrn = (Xhead); \
        while ((Xnode = Xrn)) { \
	    if (1)

#define PATRICIA_WALK_BREAK { \
	    if (Xsp != Xstack) { \
		Xrn = *(--Xsp); \
	     } else { \
		Xrn = (patricia_node_t *) 0; \
	    } \
	    continue; }

#define PATRICIA_WALK_END \
            if (Xrn->l) { \
                if (Xrn->r) { \
                    *Xsp++ = Xrn->r; \
                } \
                Xrn = Xrn->l; \
            } else if (Xrn->r) { \
                Xrn = Xrn->r; \
            } else if (Xsp != Xstack) { \
                Xrn = *(--Xsp); \
            } else { \
                Xrn = (patricia_node_t *) 0; \
            } \
        } \
    } while (0)

#endif /* _PATRICIA_H */


ext/rpatricia/rpatricia.c
/*
 *   rpatricia.c - Ruby wrapper of Net::Patricia
 *   Tatsuya Mori <mori.tatsuya@gmail.com>
 */

#include "ruby.h"
#include <stdlib.h>
#include "patricia.h"
#include <assert.h>

size_t patricia_walk_inorder(patricia_node_t *node, void_fn_t func);
char * prefix_toa2x(prefix_t *prefix, char *buff, int with_len);
char * prefix_toa2(prefix_t *prefix, char *buff);

static VALUE cPatricia, cNode;
static VALUE sym_AF_INET, sym_AF_INET6;

static void dummy(void) {}

static VALUE
p_destroy (VALUE self)
{
  patricia_tree_t *tree;

  Data_Get_Struct(self, patricia_tree_t, tree);
  Clear_Patricia(tree, dummy);
  tree->head = NULL; /* Clear_Patricia() should do this, actually */

  return Qtrue;
}

static void
p_node_mark (void *ptr)
{
  patricia_node_t *node = ptr;

  rb_gc_mark((VALUE)node->data);
}

static VALUE
wrap_node(patricia_node_t *orig)
{
  patricia_node_t *node = ALLOC(patricia_node_t);

  memcpy(node, orig, sizeof(patricia_node_t));

  return Data_Wrap_Struct(cNode, p_node_mark, -1, node);
}

static void
my_ascii2prefix(patricia_tree_t *tree, VALUE str, prefix_t *prefix)
{
  char *cstr = StringValueCStr(str);
  prefix_t *ok = ascii2prefix(cstr, prefix);

  if (!ok)
    rb_raise(rb_eArgError, "invalid prefix: %s", cstr);

  if (prefix->bitlen > tree->maxbits)
    rb_raise(rb_eArgError, "prefix length (%u) larger than maxbits (%u)",
             prefix->bitlen, tree->maxbits);
}

static VALUE
p_add (int argc, VALUE *argv, VALUE self)
{
  VALUE user_data;
  patricia_tree_t *tree;
  patricia_node_t *node;
  prefix_t prefix;

  if (argc > 2 || argc < 1) 
    return Qnil;

  Data_Get_Struct(self, patricia_tree_t, tree);
  my_ascii2prefix(tree, argv[0], &prefix);
  node = patricia_lookup(tree, &prefix);

  if (argc == 2) {
    user_data = argv[1];

    /* for backwards compatibility, we always dup and return new strings */
    if (TYPE(user_data) == T_STRING)
      user_data = rb_str_dup(user_data);
  } else {
    user_data = rb_str_new(NULL, 0);
  }
  PATRICIA_DATA_SET(node, user_data);

  /* node will be freed when parent is freed */
  return wrap_node(node);
}

static VALUE
p_remove (VALUE self, VALUE r_key)
{
  patricia_tree_t *tree;
  patricia_node_t *node;
  prefix_t prefix;

  Data_Get_Struct(self, patricia_tree_t, tree);
  my_ascii2prefix(tree, r_key, &prefix);
  node = patricia_search_exact(tree, &prefix);

  if (node) {
    patricia_remove (tree, node);
    return Qtrue;
  } else {
    return Qfalse;
  }
}

static VALUE
p_match (VALUE self, VALUE r_key)
{
  patricia_tree_t *tree;
  patricia_node_t *node;
  prefix_t prefix;
  
  Data_Get_Struct(self, patricia_tree_t, tree);
  my_ascii2prefix(tree, r_key, &prefix);
  node = patricia_search_best(tree, &prefix);

  return node ? wrap_node(node) : Qnil;
}

static VALUE
p_include (VALUE self, VALUE r_key)
{
  patricia_tree_t *tree;
  patricia_node_t *node;
  prefix_t prefix;

  Data_Get_Struct(self, patricia_tree_t, tree);
  my_ascii2prefix(tree, r_key, &prefix);
  node = patricia_search_best(tree, &prefix);

  return node ? Qtrue : Qfalse;
}

static VALUE
p_match_exact (VALUE self, VALUE r_key)
{
  patricia_tree_t *tree;
  patricia_node_t *node;
  prefix_t prefix;

  Data_Get_Struct(self, patricia_tree_t, tree);
  my_ascii2prefix(tree, r_key, &prefix);
  node = patricia_search_exact(tree, &prefix);

  return node ? wrap_node(node) : Qnil;
}

static VALUE
p_num_nodes (VALUE self)
{
  int n;
  patricia_tree_t *tree;

  Data_Get_Struct(self, patricia_tree_t, tree);
  n = tree->head ? patricia_walk_inorder(tree->head, dummy) : 0;

  return INT2NUM(n);
}

/* needed for Ruby 1.8.6, in 1.8.7 and later */
#ifndef HAVE_RB_STR_SET_LEN
static void
rb_str_set_len(VALUE str, long len)
{
  RSTRING(str)->len = len;
  RSTRING(str)->ptr[len] = '\0';
}
#endif

static VALUE
p_print_nodes (int argc, VALUE *argv, VALUE self)
{
  ID id_printf = rb_intern("printf");
  VALUE fmt = rb_str_new2("node: %s\n");
  VALUE buf = rb_str_new(0, 0);
  char *cbuf;
  patricia_tree_t *tree;
  patricia_node_t *node;
  VALUE out;
  Data_Get_Struct(self, patricia_tree_t, tree);

  rb_scan_args(argc, argv, "01", &out);
  if (NIL_P(out))
    out = rb_stdout;

  if (tree->head) {
    PATRICIA_WALK(tree->head, node) {
      rb_str_resize(buf, PATRICIA_MAXSTRLEN);
      cbuf = RSTRING_PTR(buf);
      prefix_toa2x(node->prefix, cbuf, 1);
      rb_str_set_len(buf, strlen(cbuf));
      rb_funcall(out, id_printf, 2, fmt, buf);
    } PATRICIA_WALK_END;
  }
  return Qtrue;
}

static VALUE
p_each(VALUE self)
{
  patricia_tree_t *tree;
  patricia_node_t *node;
  Data_Get_Struct(self, patricia_tree_t, tree);

  if (tree->head) {
    PATRICIA_WALK(tree->head, node) {
      rb_yield(wrap_node(node));
    } PATRICIA_WALK_END;
  } else {
    return Qnil;
  }
}

static VALUE
p_data (VALUE self)
{
  VALUE user_data;
  patricia_node_t *node;
  Data_Get_Struct(self, patricia_node_t, node);

  user_data = (VALUE)node->data;

  /* for backwards compatibility, we always dup and return new strings */
  if (TYPE(user_data) == T_STRING)
    user_data = rb_str_dup(user_data);

  return user_data;
}

static VALUE
p_network (VALUE self)
{
  patricia_node_t *node;
  VALUE str = rb_str_new(0, PATRICIA_MAXSTRLEN);
  char *cstr = RSTRING_PTR(str);

  Data_Get_Struct(self, patricia_node_t, node);
  prefix_toa2x(node->prefix, cstr, 1);
  rb_str_set_len(str, strlen(cstr));

  return str;
}

static VALUE
p_prefix (VALUE self)
{
  patricia_node_t *node;
  VALUE str = rb_str_new(0, INET6_ADDRSTRLEN);
  char *cstr = RSTRING_PTR(str);

  Data_Get_Struct(self, patricia_node_t, node);
  prefix_toa2(node->prefix, cstr);
  rb_str_set_len(str, strlen(cstr));

  return str;
}

static VALUE
p_prefixlen (VALUE self)
{
  patricia_node_t *node;
  Data_Get_Struct(self, patricia_node_t, node);
  return INT2NUM(node->prefix->bitlen);
}

/* called during GC for each node->data attached to a Patricia tree */
static void
p_tree_mark_each(prefix_t *prefix, void *data)
{
  VALUE user_data = (VALUE)data;

  rb_gc_mark(user_data);
}

static void
p_tree_mark (void *ptr)
{
  patricia_tree_t *tree = ptr;

  if (tree)
    patricia_process(tree, p_tree_mark_each);
}

static void
p_tree_free (void *ptr)
{
  patricia_tree_t *tree = ptr;
  /* no need to explicitly free each node->data, GC will do it for us */
  if (tree)
    Destroy_Patricia(tree, NULL);
}

static VALUE
p_alloc(VALUE klass)
{
  return Data_Wrap_Struct(klass, p_tree_mark, p_tree_free, NULL);
}

static VALUE
p_init(int argc, VALUE *argv, VALUE self)
{
  VALUE family;
  int maxbits;

  rb_scan_args(argc, argv, "01", &family);

  if (NIL_P(family) || family == sym_AF_INET)
    maxbits = 32;
  else if (family == sym_AF_INET6)
    maxbits = 128;
  else
    rb_raise(rb_eArgError, "unknown family (must be :AF_INET6 or :AF_INET)");

  DATA_PTR(self) = New_Patricia(maxbits);
  return self;
}

static VALUE
p_init_copy(VALUE self, VALUE orig)
{
  patricia_tree_t *orig_tree;

  Data_Get_Struct(orig, patricia_tree_t, orig_tree);
  if (orig_tree->head) {
    patricia_tree_t *tree;
    patricia_node_t *orig_node, *node;
    VALUE user_data;

    DATA_PTR(self) = tree = New_Patricia(orig_tree->maxbits);

    PATRICIA_WALK(orig_tree->head, orig_node) {
      node = patricia_lookup(tree, orig_node->prefix);
      assert(node->prefix == orig_node->prefix);

      user_data = (VALUE)(orig_node->data);
      if (T_STRING == TYPE(user_data))
        user_data = rb_str_dup(user_data);
      PATRICIA_DATA_SET(node, user_data);
    } PATRICIA_WALK_END;
  }

  return self;
}

static VALUE
p_family(VALUE self)
{
  patricia_tree_t *tree;

  Data_Get_Struct(self, patricia_tree_t, tree);

  switch (tree->maxbits) {
  case 32: return sym_AF_INET;
  case 128: return sym_AF_INET6;
  }
  assert(0 && "unknown maxbits, corrupt tree");
  return Qnil;
}

void
Init_rpatricia (void)
{
  cPatricia = rb_define_class("Patricia", rb_cObject);
  rb_include_module(cPatricia, rb_mEnumerable);
  cNode = rb_define_class_under(cPatricia, "Node", rb_cObject);
  sym_AF_INET = ID2SYM(rb_intern("AF_INET"));
  sym_AF_INET6 = ID2SYM(rb_intern("AF_INET6"));

  /* allocate new Patricia object, called before initialize  */
  rb_define_alloc_func(cPatricia, p_alloc);
  rb_define_private_method(cPatricia, "initialize", p_init, -1);
  rb_define_method(cPatricia, "initialize_copy", p_init_copy, 1);

  /*---------- methods to tree ----------*/
  /* add string */
  rb_define_method(cPatricia, "add", p_add, -1);
  rb_define_method(cPatricia, "add_node", p_add, -1);
  rb_define_method(cPatricia, "family", p_family, 0);

  /* match prefix */
  rb_define_method(cPatricia, "match_best", p_match, 1); 
  rb_define_method(cPatricia, "search_best", p_match, 1); 

  /* exact match  */
  rb_define_method(cPatricia, "match_exact", p_match_exact, 1); 
  rb_define_method(cPatricia, "search_exact", p_match_exact, 1); 

  /* check existence */
  rb_define_method(cPatricia, "include?", p_include, 1);

  /* removal */
  rb_define_method(cPatricia, "remove", p_remove, 1); 
  rb_define_method(cPatricia, "remove_node", p_remove, 1); 

  /* derivatives of climb */
  rb_define_method(cPatricia, "num_nodes", p_num_nodes, 0);
  rb_define_method(cPatricia, "show_nodes", p_print_nodes, -1);
  rb_define_method(cPatricia, "each", p_each, 0);

  /* destroy tree */
  rb_define_method(cPatricia, "destroy", p_destroy, 0); 
  rb_define_method(cPatricia, "clear", p_destroy, 0); 

  /*---------- methods to node ----------*/
  rb_define_method(cNode, "data", p_data, 0);
  rb_define_method(cNode, "show_data", p_data, 0);
  rb_define_method(cNode, "network", p_network, 0);
  rb_define_method(cNode, "prefix", p_prefix, 0);
  rb_define_method(cNode, "prefixlen", p_prefixlen, 0);
}
